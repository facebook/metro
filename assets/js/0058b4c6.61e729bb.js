"use strict";(self.webpackChunkmetro_website=self.webpackChunkmetro_website||[]).push([[849],{86164:e=>{e.exports=JSON.parse('{"version":{"pluginId":"default","version":"current","label":"Next","banner":null,"badge":false,"noIndex":false,"className":"docs-version-current","isLast":true,"docsSidebars":{"docs":[{"type":"category","label":"Introduction","items":[{"type":"link","label":"Getting Started","href":"/docs/getting-started","docId":"getting-started","unlisted":false},{"type":"link","label":"Concepts","href":"/docs/concepts","docId":"concepts","unlisted":false}],"collapsed":true,"collapsible":true},{"type":"category","label":"API Reference","items":[{"type":"link","label":"Bundling API","href":"/docs/api","docId":"api","unlisted":false},{"type":"link","label":"Module API","href":"/docs/module-api","docId":"module-api","unlisted":false},{"type":"link","label":"Configuring Metro","href":"/docs/configuration","docId":"configuration","unlisted":false},{"type":"link","label":"Metro CLI Options","href":"/docs/cli","docId":"cli","unlisted":false}],"collapsed":true,"collapsible":true},{"type":"category","label":"Guides","items":[{"type":"link","label":"Package Exports Support (New)","href":"/docs/package-exports","docId":"package-exports","unlisted":false},{"type":"link","label":"Troubleshooting","href":"/docs/troubleshooting","docId":"troubleshooting","unlisted":false}],"collapsed":true,"collapsible":true},{"type":"category","label":"Contributing","items":[{"type":"link","label":"Local Development Setup","href":"/docs/local-development","docId":"local-development","unlisted":false}],"collapsed":true,"collapsible":true},{"type":"category","label":"Deep Dives","items":[{"type":"link","label":"Bundle Formats","href":"/docs/bundling","docId":"bundling","unlisted":false},{"type":"link","label":"Caching","href":"/docs/caching","docId":"caching","unlisted":false},{"type":"link","label":"Module Resolution","href":"/docs/resolution","docId":"resolution","unlisted":false},{"type":"link","label":"Source Map Format","href":"/docs/source-map-format","docId":"source-map-format","unlisted":false}],"collapsed":true,"collapsible":true}]},"docs":{"api":{"id":"api","title":"Bundling API","description":"Quick Start","sidebar":"docs"},"bundling":{"id":"bundling","title":"Bundle Formats","description":"When bundling, each of the modules gets assigned a numeric id, meaning no dynamic requires are supported. Requires are changed by its numeric version, and modules are stored in different possible formats. Three different formats of bundling are supported:","sidebar":"docs"},"caching":{"id":"caching","title":"Caching","description":"Out of the box, Metro speeds up builds using a local cache of transformed modules. Thanks to this cache, Metro doesn\'t need to retransform modules unless the source code (or current configuration) has changed since the last time they were transformed.","sidebar":"docs"},"cli":{"id":"cli","title":"Metro CLI Options","description":"The metro command line runner has a number of useful options. You can run `metro","sidebar":"docs"},"concepts":{"id":"concepts","title":"Concepts","description":"Metro is a JavaScript bundler. It takes in an entry file and various options, and gives you back a single JavaScript file that includes all your code and its dependencies.","sidebar":"docs"},"configuration":{"id":"configuration","title":"Configuring Metro","description":"A Metro config can be created in these three ways (ordered by priority):","sidebar":"docs"},"getting-started":{"id":"getting-started","title":"Getting Started","description":"Install Metro using npm:","sidebar":"docs"},"local-development":{"id":"local-development","title":"Local Development Setup","description":"This page includes tips for developers working on Metro itself, including how to test your changes within other local projects.","sidebar":"docs"},"module-api":{"id":"module-api","title":"Module API","description":"Metro is designed to allow code written for Node (or for bundlers targeting the Web) to run mostly unmodified. The main APIs available to application code are listed below.","sidebar":"docs"},"package-exports":{"id":"package-exports","title":"Package Exports Support (New)","description":"Background","sidebar":"docs"},"resolution":{"id":"resolution","title":"Module Resolution","description":"Module resolution is the process of translating module names to module paths at build time. For example, if your project contains the code:","sidebar":"docs"},"source-map-format":{"id":"source-map-format","title":"Source Map Format","description":"Metro produces standard source maps along with its JavaScript bundle output. In addition to the standard information, Metro encodes extra information in vendor-specific fields within the source map. This page serves as a specification for this encoding.","sidebar":"docs"},"troubleshooting":{"id":"troubleshooting","title":"Troubleshooting","description":"Uh oh, something went wrong? Use this guide to resolve issues with Metro.","sidebar":"docs"}}}}')}}]);