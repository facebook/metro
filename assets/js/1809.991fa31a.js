"use strict";(self.webpackChunkmetro_website=self.webpackChunkmetro_website||[]).push([[1809],{1809:(e,s,t)=>{t.r(s)}}]);