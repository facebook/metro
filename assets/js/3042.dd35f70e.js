"use strict";(self.webpackChunkmetro_website=self.webpackChunkmetro_website||[]).push([[3042],{65932:(e,t,i)=>{i.d(t,{A:()=>o});i(96540);var n=i(34164),a=i(50539),r=i(9303),s=i(74848);function o(e){var t=e.className;return(0,s.jsx)("main",{className:(0,n.A)("container margin-vert--xl",t),children:(0,s.jsx)("div",{className:"row",children:(0,s.jsxs)("div",{className:"col col--6 col--offset-3",children:[(0,s.jsx)(r.A,{as:"h1",className:"hero__title",children:(0,s.jsx)(a.default,{id:"theme.NotFound.title",description:"The title of the 404 page",children:"Page Not Found"})}),(0,s.jsx)("p",{children:(0,s.jsx)(a.default,{id:"theme.NotFound.p1",description:"The first paragraph of the 404 page",children:"We could not find what you were looking for."})}),(0,s.jsx)("p",{children:(0,s.jsx)(a.default,{id:"theme.NotFound.p2",description:"The 2nd paragraph of the 404 page",children:"Please contact the owner of the site that linked you to the original URL and let them know their link is broken."})})]})})})}},73042:(e,t,i)=>{i.r(t),i.d(t,{default:()=>l});i(96540);var n=i(50539),a=i(59144),r=i(51986),s=i(65932),o=i(74848);function l(){var e=(0,n.translate)({id:"theme.NotFound.title",message:"Page Not Found"});return(0,o.jsxs)(o.Fragment,{children:[(0,o.jsx)(a.be,{title:e}),(0,o.jsx)(r.A,{children:(0,o.jsx)(s.A,{})})]})}}}]);