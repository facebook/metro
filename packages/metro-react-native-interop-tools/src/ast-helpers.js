/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+metro_bundler
 * @flow strict-local
 * @format
 */

import type {
  File as BabelNodeFile,
  FunctionTypeAnnotation as BabelNodeFunctionTypeAnnotation,
  FunctionTypeParam as BabelNodeFunctionTypeParam,
  ObjectTypeAnnotation as BabelNodeObjectTypeAnnotation,
  ObjectTypeProperty as BabelNodeObjectTypeProperty,
  ObjectTypeSpreadProperty as BabelNodeObjectTypeSpreadProperty,
  TupleTypeAnnotation as BabelNodeTupleTypeAnnotation,
  NullableTypeAnnotation as BabelNodeNullableTypeAnnotation,
  GenericTypeAnnotation as BabelNodeGenericTypeAnnotation,
  UnionTypeAnnotation as BabelNodeUnionTypeAnnotation,
  IntersectionTypeAnnotation as BabelNodeIntersectionTypeAnnotation,
  ArrayTypeAnnotation as BabelNodeArrayTypeAnnotation,
  InterfaceExtends as BabelNodeInterfaceExtends,
  StringLiteralTypeAnnotation as BabelNodeStringLiteralTypeAnnotation,
  NumberLiteralTypeAnnotation as BabelNodeNumberLiteralTypeAnnotation,
  InterfaceDeclaration as BabelNodeInterfaceDeclaration,
  FlowType as BabelNodeFlowType,
  Statement as BabelNodeStatement,
} from '@babel/types';

import type {
  InterfaceExtends,
  AnyTypeAnnotation,
  ArrayTypeAnnotation,
  NullableTypeAnnotation,
  FunctionTypeAnnotation,
  ObjectTypeAnnotation,
  FunctionTypeParam,
  ObjectTypeProperty,
  TupleTypeAnnotation,
  GenericTypeAnnotation,
  UnionTypeAnnotation,
  IntersectionTypeAnnotation,
  StringLiteralTypeAnnotation,
  NumberLiteralTypeAnnotation,
} from './type-annotation.js';

export type BoundarySchema = {
  typegenSchema: {},
  source: string,
  ...
};

export function isTurboModule(i: BabelNodeInterfaceExtends): boolean {
  return (
    i.id.name === 'TurboModule' &&
    (i.typeParameters == null || i.typeParameters.params.length === 0)
  );
}

export function getNodeLoc(
  loc: ?BabelNodeSourceLocation,
): ?BabelNodeSourceLocation {
  return loc == null
    ? null
    : {
        start: loc.start,
        end: loc.end,
      };
}

export function getTypeAnnotation(typeNode: BabelNodeFlow): AnyTypeAnnotation {
  switch (typeNode.type) {
    case 'BooleanTypeAnnotation':
    case 'NumberTypeAnnotation':
    case 'StringTypeAnnotation':
    case 'VoidTypeAnnotation':
    case 'AnyTypeAnnotation':
      return {
        type: typeNode.type,
        loc: getNodeLoc(typeNode.loc),
      };

    case 'NumberLiteralTypeAnnotation':
      return getNumberLiteralTypeAnnotation(typeNode);

    case 'StringLiteralTypeAnnotation':
      return getStringLiteralTypeAnnotation(typeNode);

    case 'ArrayTypeAnnotation':
      return getArrayTypeAnnotation(typeNode);

    case 'NullableTypeAnnotation':
      return getNullableTypeAnnotation(typeNode);

    case 'FunctionTypeAnnotation':
      return getFunctionTypeAnnotation(typeNode);

    case 'ObjectTypeAnnotation':
      return getObjectTypeAnnotation(typeNode);

    case 'TupleTypeAnnotation':
      return getTupleTypeAnnotation(typeNode);

    case 'GenericTypeAnnotation':
      return getGenericTypeAnnotation(typeNode);

    case 'UnionTypeAnnotation':
      return getUnionTypeAnnotation(typeNode);

    case 'IntersectionTypeAnnotation':
      return getIntersectionTypeAnnotation(typeNode);

    default:
      return {type: 'UnknownTypeAnnotation', loc: null};
  }
}

export function getFunctionTypeAnnotation(
  typeNode: BabelNodeFunctionTypeAnnotation,
): FunctionTypeAnnotation {
  return {
    type: 'FunctionTypeAnnotation',
    loc: getNodeLoc(typeNode.loc),
    params: typeNode.params.map(getFunctionTypeParameter),
    returnTypeAnnotation: getTypeAnnotation(typeNode.returnType),
  };
}

export function getFunctionTypeParameter(
  param: BabelNodeFunctionTypeParam,
): FunctionTypeParam {
  return {
    name: param.name?.name,
    typeAnnotation: getTypeAnnotation(param.typeAnnotation),
  };
}

export function getObjectTypeAnnotation(
  typeNode: BabelNodeObjectTypeAnnotation,
): ObjectTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    properties: typeNode.properties.map(
      (
        property:
          | BabelNodeObjectTypeProperty
          | BabelNodeObjectTypeSpreadProperty,
      ) => {
        return property.type === 'ObjectTypeProperty'
          ? getObjectTypeProperty(property)
          : getObjectTypeSpreadProperty(property);
      },
    ),
  };
}

//TODO T127639272 add support for spread properties
export function getObjectTypeSpreadProperty(
  typeProperty: BabelNodeObjectTypeSpreadProperty,
): ObjectTypeProperty {
  return {
    loc: getNodeLoc(typeProperty.loc),
    name: '',
    optional: false,
    typeAnnotation: {
      type: 'UnknownTypeAnnotation',
      loc: null,
    },
  };
}

export function getObjectTypeProperty(
  typeProperty: BabelNodeObjectTypeProperty,
): ObjectTypeProperty {
  return {
    loc: getNodeLoc(typeProperty.loc),
    name: getNameFromID(typeProperty.key),
    optional: typeProperty.optional,
    typeAnnotation: getTypeAnnotation(typeProperty.value),
  };
}

export function getTupleTypeAnnotation(
  typeNode: BabelNodeTupleTypeAnnotation,
): TupleTypeAnnotation {
  return {
    type: 'TupleTypeAnnotation',
    loc: getNodeLoc(typeNode.loc),
    types: typeNode.types.map(getTypeAnnotation),
  };
}

export function getNullableTypeAnnotation(
  typeNode: BabelNodeNullableTypeAnnotation,
): NullableTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    typeAnnotation: getTypeAnnotation(typeNode.typeAnnotation),
  };
}

// Flow can check for exhaustive switch cases where eslint can not, using flow to cover all possible cases is better than having default fallbacks.
// eslint-disable-next-line consistent-return
export function getNameFromID(
  node:
    | BabelNodeIdentifier
    | BabelNodeQualifiedTypeIdentifier
    | BabelNodeStringLiteral,
): string {
  switch (node.type) {
    case 'QualifiedTypeIdentifier':
      return node.id.name;
    case 'StringLiteral':
      return node.value;
    case 'Identifier':
      return node.name;
  }
}

export function getGenericTypeAnnotation(
  typeNode: BabelNodeGenericTypeAnnotation,
): GenericTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    name: getNameFromID(typeNode.id),
    typeParameters: typeNode.typeParameters?.params
      ? typeNode.typeParameters.params?.map(getTypeAnnotation)
      : [],
  };
}

export function getUnionTypeAnnotation(
  typeNode: BabelNodeUnionTypeAnnotation,
): UnionTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    types: typeNode.types.map(getTypeAnnotation),
  };
}

export function getIntersectionTypeAnnotation(
  typeNode: BabelNodeIntersectionTypeAnnotation,
): IntersectionTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    types: typeNode.types.map(getTypeAnnotation),
  };
}

export function getArrayTypeAnnotation(
  typeNode: BabelNodeArrayTypeAnnotation,
): ArrayTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    elementType: getTypeAnnotation(typeNode.elementType),
  };
}

export function getStringLiteralTypeAnnotation(
  typeNode: BabelNodeStringLiteralTypeAnnotation,
): StringLiteralTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    value: typeNode.value,
  };
}

export function getNumberLiteralTypeAnnotation(
  typeNode: BabelNodeNumberLiteralTypeAnnotation,
): NumberLiteralTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    value: typeNode.value,
  };
}

export function getInterfaceExtends(
  interfaceNode: BabelNodeInterfaceExtends,
): InterfaceExtends {
  return {
    loc: getNodeLoc(interfaceNode.loc),
    name: getNameFromID(interfaceNode.id),
    typeParameters: getTypeParameters(interfaceNode.typeParameters?.params),
  };
}

function interfaceDeclarationReducer(
  interfaceDeclaration: ?BabelNodeInterfaceDeclaration,
  bodyNode: BabelNodeStatement,
): ?BabelNodeInterfaceDeclaration {
  if (
    bodyNode.type === 'ExportNamedDeclaration' &&
    bodyNode.declaration != null
  ) {
    return interfaceDeclarationReducer(
      interfaceDeclaration,
      bodyNode.declaration,
    );
  } else if (
    bodyNode.type === 'InterfaceDeclaration' &&
    bodyNode.extends?.some(isTurboModule)
  ) {
    return bodyNode;
  }
  return interfaceDeclaration;
}

export function getTypeParameters(
  params: ?Array<BabelNodeFlowType>,
): Array<AnyTypeAnnotation> {
  return params?.map(getTypeAnnotation) ?? [];
}

export function getBoundarySchemaFromAST(
  ast: BabelNodeFile,
  source: string,
): BoundarySchema {
  const schema: BoundarySchema = {
    typegenSchema: {},
    source,
  };
  const interfaceNode: ?BabelNodeInterfaceDeclaration = ast.program.body.reduce(
    interfaceDeclarationReducer,
    null,
  );
  if (interfaceNode != null) {
    schema.typegenSchema[interfaceNode.id.name] = {
      typeAnnotation: {
        type: 'InterfaceDeclarationTypeAnnotation',
        innerType: getTypeAnnotation(interfaceNode.body),
        extends: interfaceNode.extends?.map(getInterfaceExtends),
      },
    };
  }
  return schema;
}
