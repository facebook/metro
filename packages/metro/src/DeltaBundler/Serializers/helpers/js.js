/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 * @oncall react_native
 */

'use strict';

import type {MixedOutput, Module} from '../../types.flow';
import type {JsOutput} from 'metro-transform-worker';

const invariant = require('invariant');
const {addParamsToDefineCall} = require('metro-transform-plugins');
const path = require('path');

export type Options = $ReadOnly<{
  createModuleId: string => number | string,
  dev: boolean,
  includeAsyncPaths: boolean,
  projectRoot: string,
  serverRoot: string,
  sourceUrl: ?string,
  ...
}>;

function wrapModule(module: Module<>, options: Options): string {
  const output = getJsOutput(module);

  if (output.type.startsWith('js/script')) {
    return output.data.code;
  }

  const params = getModuleParams(module, options);
  return addParamsToDefineCall(output.data.code, ...params);
}

function getModuleParams(module: Module<>, options: Options): Array<mixed> {
  const moduleId = options.createModuleId(module.path);

  const paths: {[moduleID: number | string]: mixed} = {};
  let hasPaths = false;
  const dependencyMapArray = Array.from(module.dependencies.values()).map(
    dependency => {
      const id = options.createModuleId(dependency.absolutePath);
      if (options.includeAsyncPaths && dependency.data.data.asyncType != null) {
        hasPaths = true;
        invariant(
          options.sourceUrl != null,
          'sourceUrl is required when includeAsyncPaths is true',
        );

        // TODO: Only include path if the target is not in the bundle

        // Construct a server-relative URL for the split bundle, propagating
        // most parameters from the main bundle's URL.

        const {searchParams} = new URL(options.sourceUrl);
        searchParams.set('modulesOnly', 'true');
        searchParams.set('runModule', 'false');

        const bundlePath = path.relative(
          options.serverRoot,
          dependency.absolutePath,
        );
        paths[id] =
          '/' +
          path.join(
            path.dirname(bundlePath),
            // Strip the file extension
            path.basename(bundlePath, path.extname(bundlePath)),
          ) +
          '.bundle?' +
          searchParams.toString();
      }
      return id;
    },
  );

  const params = [
    moduleId,
    hasPaths
      ? {
          // $FlowIgnore[not-an-object] Intentionally spreading an array into an object
          ...dependencyMapArray,
          paths,
        }
      : dependencyMapArray,
  ];

  if (options.dev) {
    // Add the relative path of the module to make debugging easier.
    // This is mapped to `module.verboseName` in `require.js`.
    params.push(path.relative(options.projectRoot, module.path));
  }

  return params;
}

function getJsOutput(
  module: $ReadOnly<{
    output: $ReadOnlyArray<MixedOutput>,
    path?: string,
    ...
  }>,
): JsOutput {
  const jsModules = module.output.filter(({type}) => type.startsWith('js/'));

  invariant(
    jsModules.length === 1,
    `Modules must have exactly one JS output, but ${
      module.path ?? 'unknown module'
    } has ${jsModules.length} JS outputs.`,
  );

  const jsOutput: JsOutput = (jsModules[0]: any);

  invariant(
    Number.isFinite(jsOutput.data.lineCount),
    `JS output must populate lineCount, but ${
      module.path ?? 'unknown module'
    } has ${jsOutput.type} output with lineCount '${jsOutput.data.lineCount}'`,
  );

  return jsOutput;
}

function isJsModule(module: Module<>): boolean {
  return module.output.filter(isJsOutput).length > 0;
}

function isJsOutput(output: MixedOutput): boolean %checks {
  return output.type.startsWith('js/');
}

module.exports = {
  getJsOutput,
  getModuleParams,
  isJsModule,
  wrapModule,
};
