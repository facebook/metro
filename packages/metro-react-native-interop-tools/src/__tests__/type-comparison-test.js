/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 * @oncall react_native
 */

import {parse} from 'hermes-parser';
import {getTypeAnnotation} from '../ast-helpers.js';
import {compareTypeAnnotation} from '../type-comparison.js';
import type {AnyTypeAnnotation} from '../type-annotation.js';
import type {FlowType, Statement} from '@babel/types';

function getTypeFromAlias(body: Statement): FlowType {
  if (body.type !== 'TypeAlias') {
    throw new Error('This function support only TypeAlias');
  }
  return body.right;
}

function getTypeFromCode(stringType: string): AnyTypeAnnotation {
  const ast = parse('type T =' + stringType, {
    babel: true,
    sourceType: 'module',
    sourceFilename: 'NativeDeviceManager.js',
  });
  const astType = getTypeFromAlias(ast.program.body[0]);
  return getTypeAnnotation(astType);
}

test.each([
  ['boolean', 'boolean'],
  ['string', 'string'],
  ['number', 'number'],
  ['void', 'void'],
  ['boolean', 'number'],
  ["'a'", "'a'"],
  ["'a'", "'b'"],
  ['8', '8'],
  ['2', '8'],
  ['2', 'number'],
  ['number', '2'],
  ['string', "'a'"],
  ["'a'", 'string'],
  ['true', 'true'],
  ['true', 'false'],
  ['string', '?string'],
  ['?number', 'number'],
  ['?boolean', 'true'],
  ['?string', '?string'],
  ['?string', "?'foo'"],
  ['?string', '?number'],
  ['null', 'null'],
  ['?string', 'null'],
  ['?boolean', 'void'],
  ['() => boolean', '() => ?boolean'],
  ['() => ?boolean', '() => boolean'],
  ['() => true', '() => boolean'],
  ['(test: ?boolean) => true', '(test: boolean) => true'],
  ['(test?: string) => void', '() => void'],
  ['(test: string) => void', '() => void'],
  ['() => void', '(test?: string) => void'],
  ['() => void', '(test?: string, test2: number) => void'],
  ['(test?: boolean) => true', '(test?: string) => true'],
  ['(test: string) => ?true', '() => void'],
  ['{name: string, age: ?number }', '{name: string, age: number }'],
  ['{name: string, age: number }', '{name: string, age?: number }'],
  ['{name: string, age: number }', '{name: string}'],
  ['{name: string, age?: number }', '{name: string}'],
  ['{name: string}', '{name: string, ...}'],
  ['{name: string}', '{name: string, age: number}'],
  ['{name: string}', '{name: string, age?: number}'],
  ['() => {name: string, age?: number }', '() => {name: string}'],
  ['() => {name: string, age: number }', '() => {name: string}'],
  ['() => {name: string}', '() => {name: string, age: number}'],
  ['() => {name: string}', '() => {name: string, age: ?number}'],
])('comparing basic types', (left, right) => {
  const result = compareTypeAnnotation(
    getTypeFromCode(left),
    getTypeFromCode(right),
    false,
  );
  let messages: string = '';
  result.forEach(error => {
    messages = messages + error.message + '\n  \t\t';
  });
  messages = messages === '' ? 'no errors' : messages;
  expect(`
    left-type:
      ${left}
    right-type:
      ${right}
    output:
      ${messages}`).toMatchSnapshot();
});
