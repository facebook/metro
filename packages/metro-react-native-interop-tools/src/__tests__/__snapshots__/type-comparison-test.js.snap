// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comparing basic types 1`] = `
"
    left-type:
      boolean
    right-type:
      boolean
    output:
      no errors"
`;

exports[`comparing basic types 2`] = `
"
    left-type:
      string
    right-type:
      string
    output:
      no errors"
`;

exports[`comparing basic types 3`] = `
"
    left-type:
      number
    right-type:
      number
    output:
      no errors"
`;

exports[`comparing basic types 4`] = `
"
    left-type:
      void
    right-type:
      void
    output:
      no errors"
`;

exports[`comparing basic types 5`] = `
"
    left-type:
      boolean
    right-type:
      number
    output:
      Error: cannot change boolean to number because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 6`] = `
"
    left-type:
      'a'
    right-type:
      'a'
    output:
      no errors"
`;

exports[`comparing basic types 7`] = `
"
    left-type:
      'a'
    right-type:
      'b'
    output:
      Error: cannot change string literal with value 'a' to string literal with value 'b' because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 8`] = `
"
    left-type:
      8
    right-type:
      8
    output:
      no errors"
`;

exports[`comparing basic types 9`] = `
"
    left-type:
      2
    right-type:
      8
    output:
      Error: cannot change number literal with value '2' to number literal with value '8' because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 10`] = `
"
    left-type:
      2
    right-type:
      number
    output:
      Error: cannot change number literal to number because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 11`] = `
"
    left-type:
      number
    right-type:
      2
    output:
      no errors"
`;

exports[`comparing basic types 12`] = `
"
    left-type:
      string
    right-type:
      'a'
    output:
      no errors"
`;

exports[`comparing basic types 13`] = `
"
    left-type:
      'a'
    right-type:
      string
    output:
      Error: cannot change string literal to string because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 14`] = `
"
    left-type:
      true
    right-type:
      true
    output:
      no errors"
`;

exports[`comparing basic types 15`] = `
"
    left-type:
      true
    right-type:
      false
    output:
      Error: cannot change boolean literal with value 'true' to boolean literal with value 'false' because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 16`] = `
"
    left-type:
      string
    right-type:
      ?string
    output:
      Error: cannot change string to nullable string because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 17`] = `
"
    left-type:
      ?number
    right-type:
      number
    output:
      no errors"
`;

exports[`comparing basic types 18`] = `
"
    left-type:
      ?boolean
    right-type:
      true
    output:
      no errors"
`;

exports[`comparing basic types 19`] = `
"
    left-type:
      ?string
    right-type:
      ?string
    output:
      no errors"
`;

exports[`comparing basic types 20`] = `
"
    left-type:
      ?string
    right-type:
      ?'foo'
    output:
      no errors"
`;

exports[`comparing basic types 21`] = `
"
    left-type:
      ?string
    right-type:
      ?number
    output:
      Error: cannot change string to number because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 22`] = `
"
    left-type:
      null
    right-type:
      null
    output:
      no errors"
`;

exports[`comparing basic types 23`] = `
"
    left-type:
      ?string
    right-type:
      null
    output:
      no errors"
`;

exports[`comparing basic types 24`] = `
"
    left-type:
      ?boolean
    right-type:
      void
    output:
      no errors"
`;

exports[`comparing basic types 25`] = `
"
    left-type:
      () => boolean
    right-type:
      () => ?boolean
    output:
      no errors"
`;

exports[`comparing basic types 26`] = `
"
    left-type:
      () => ?boolean
    right-type:
      () => boolean
    output:
      Error: cannot change nullable boolean to boolean because is incompatible with what the native code returns.
  		"
`;

exports[`comparing basic types 27`] = `
"
    left-type:
      () => true
    right-type:
      () => boolean
    output:
      no errors"
`;

exports[`comparing basic types 28`] = `
"
    left-type:
      (test: ?boolean) => true
    right-type:
      (test: boolean) => true
    output:
      no errors"
`;

exports[`comparing basic types 29`] = `
"
    left-type:
      (test?: string) => void
    right-type:
      () => void
    output:
      no errors"
`;

exports[`comparing basic types 30`] = `
"
    left-type:
      (test: string) => void
    right-type:
      () => void
    output:
      Error: cannot remove required parameter string because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 31`] = `
"
    left-type:
      () => void
    right-type:
      (test?: string) => void
    output:
      no errors"
`;

exports[`comparing basic types 32`] = `
"
    left-type:
      () => void
    right-type:
      (test?: string, test2: number) => void
    output:
      Error: cannot add new required parameter number because native will not provide it.
  		"
`;

exports[`comparing basic types 33`] = `
"
    left-type:
      (test?: boolean) => true
    right-type:
      (test?: string) => true
    output:
      Error: cannot change boolean to string because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 34`] = `
"
    left-type:
      (test: string) => ?true
    right-type:
      () => void
    output:
      Error: cannot change nullable boolean literal to void because is incompatible with what the native code returns.
  		Error: cannot remove required parameter string because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 35`] = `
"
    left-type:
      {name: string, age: ?number }
    right-type:
      {name: string, age: number }
    output:
      no errors"
`;

exports[`comparing basic types 36`] = `
"
    left-type:
      {name: string, age: number }
    right-type:
      {name: string, age?: number }
    output:
      Error: cannot change required number to optional number because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 37`] = `
"
    left-type:
      {name: string, age: number }
    right-type:
      {name: string}
    output:
      Error: cannot remove number from object properties because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 38`] = `
"
    left-type:
      {name: string, age?: number }
    right-type:
      {name: string}
    output:
      no errors"
`;

exports[`comparing basic types 39`] = `
"
    left-type:
      {name: string}
    right-type:
      {name: string, ...}
    output:
      no errors"
`;

exports[`comparing basic types 40`] = `
"
    left-type:
      {name: string}
    right-type:
      {name: string, age: number}
    output:
      Error: cannot add number to object properties because native code will break when js calls it.
  		"
`;

exports[`comparing basic types 41`] = `
"
    left-type:
      {name: string}
    right-type:
      {name: string, age?: number}
    output:
      no errors"
`;

exports[`comparing basic types 42`] = `
"
    left-type:
      () => {name: string, age?: number }
    right-type:
      () => {name: string}
    output:
      no errors"
`;

exports[`comparing basic types 43`] = `
"
    left-type:
      () => {name: string, age: number }
    right-type:
      () => {name: string}
    output:
      Error: cannot remove number from object properties because it is incompatible with what the native code returns.
  		"
`;

exports[`comparing basic types 44`] = `
"
    left-type:
      () => {name: string}
    right-type:
      () => {name: string, age: number}
    output:
      Error: cannot add number to object properties because it is incompatible with what the native code returns.
  		"
`;

exports[`comparing basic types 45`] = `
"
    left-type:
      () => {name: string}
    right-type:
      () => {name: string, age: ?number}
    output:
      Error: cannot add nullable number to object properties because it is incompatible with what the native code returns.
  		"
`;
